crates/sargon/src/factor_instances_provider/next_index_assigner/next_derivation_entity_index_assigner.rs

```diff
@@ -0,0 +1,148 @@
+use crate::prelude::*;
+
+/// An assigner of derivation entity indices, used by the FactorInstancesProvider
+/// to map `IndexAgnosticPath` -> `DerivationPath` for some FactorSource on
+/// some NetworkID.
+///
+/// This assigner works with the:
+/// * cache
+/// * profile
+/// * local offsets
+///
+/// More specifically the assigner's `next` method performs approximately this
+/// operation:
+///
+/// ```ignore
+/// pub fn next(
+///    &mut self,
+///    fs_id: FactorSourceIDFromHash,
+///    path: IndexAgnosticPath,
+/// ) -> Result<HDPathComponent> {
+///     let next_from_cache = self.cache_analyzing.next(fs_id, path).unwrap_or(0);
+///     let next_from_profile = self.profile_analyzing.next(fs_id, path).unwrap_or(0);
+///     
+///     let max_index = std::cmp::max(next_from_profile, next_from_cache);
+///     let ephemeral_offset = self.ephemeral_offsets.reserve()?;
+///
+///     max_index + ephemeral_offset
+/// ```
+pub struct NextDerivationEntityIndexAssigner {
+    profile_analyzing: NextDerivationEntityIndexProfileAnalyzingAssigner,
+    cache_analyzing: NextDerivationEntityIndexCacheAnalyzingAssigner,
+    ephemeral_offsets: NextDerivationEntityIndexWithEphemeralOffsets,
+}
+
+impl NextDerivationEntityIndexAssigner {
+    pub fn new(
+        network_id: NetworkID,
+        profile: impl Into<Option<Arc<Profile>>>,
+        cache: FactorInstancesCache,
+    ) -> Self {
+        let profile_analyzing =
+            NextDerivationEntityIndexProfileAnalyzingAssigner::new(
+                network_id, profile,
+            );
+        let cache_analyzing =
+            NextDerivationEntityIndexCacheAnalyzingAssigner::new(cache);
+        let ephemeral_offsets =
+            NextDerivationEntityIndexWithEphemeralOffsets::default();
+        Self {
+            profile_analyzing,
+            cache_analyzing,
+            ephemeral_offsets,
+        }
+    }
+
+    /// Returns the next index for the given `FactorSourceIDFromHash` and
+    /// `IndexAgnosticPath`, by analyzing the cache, the profile and adding
+    /// local ephemeral offsets.
+    pub fn next(
+        &self,
+        factor_source_id: FactorSourceIDFromHash,
+        index_agnostic_path: IndexAgnosticPath,
+    ) -> Result<HDPathComponent> {
+        let default_index = HDPathComponent::from_local_key_space(
+            0u32,
+            index_agnostic_path.key_space,
+        )?;
+
+        let maybe_next_from_cache = self
+            .cache_analyzing
+            .next(factor_source_id, index_agnostic_path)?;
```

Comment:
TODO

Reply:
Right so you need to check the impl of `next` on `NextDerivationEntityIndexCacheAnalyzingAssigner`, it returns `Ok(None)` if no value is found. The only reason it can fail is if we got an overflow (adding `1` to `MAX` value). So this feels correct, right?

Reply:
it is a theoretical situation which we dont need to care about really. The size of that Profile which has all kesy in the key space does not fit an a computer...